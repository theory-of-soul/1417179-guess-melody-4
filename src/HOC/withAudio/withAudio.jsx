import React from "react";
import PropTypes from "prop-types";

const withAudio = (Component) => {
  class WithAudio extends React.PureComponent {
    constructor(props) {
      super(props);
      this.audioPlayerRef = React.createRef();
      this.state = {
        isLoading: true,
        isPlaying: props.isPlaying
      };
    }

    componentDidMount() {
      const audioPlayer = this.audioPlayerRef.current;

      audioPlayer.oncanplaythrough = () => {
        this.setState({
          isLoading: false
        });
      };

      audioPlayer.onplay = () => {
        this.setState({
          isPlaying: true
        });
      };

      audioPlayer.onpause = () => {
        this.setState({
          isPlaying: false
        });
      };
    }

    componentDidUpdate() {
      const audioPlayer = this.audioPlayerRef.current;

      if (this.props.isPlaying) {
        audioPlayer.play();
      } else {
        audioPlayer.pause();
      }
    }

    componentWillUnmount() {
      const audioPlayer = this.audioPlayerRef.current;
      audioPlayer.oncanplaythrough = null;
      audioPlayer.onplay = null;
      audioPlayer.onpause = null;
      audioPlayer.src = ``;
    }

    render() {
      return (
        <Component
          {...this.props}
          isPlaying={this.state.isPlaying}
          isLoading={this.state.isLoading}
          handlePlay={() => {
            this.setState((prevState) => ({
              isPlaying: !prevState.isPlaying
            }));
            this.props.handlePlay();
          }}
        >
          <audio
            ref={this.audioPlayerRef}
            src={this.props.src}
          />
        </Component>
      );
    }
  }

  WithAudio.propTypes = {
    src: PropTypes.string.isRequired,
    isPlaying: PropTypes.bool.isRequired,
    handlePlay: PropTypes.func.isRequired,
  };

  return WithAudio;
};


export default withAudio;
