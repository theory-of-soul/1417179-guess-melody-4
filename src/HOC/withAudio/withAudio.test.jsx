import React from "react";
import renderer from 'react-test-renderer';
import withAudio from "./withAudio";
import PropTypes from "prop-types";

const src = ``;
const options = {
  createNodeMock: (element) => {
    if (element.type === `audio`) {
      return {
        oncanplaythrough() {},
      };
    }
    return null;
  }
};

const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

const MockComponentWrapped = withAudio(MockComponent);

describe(` Hoc withAudio snapshot tests`, () => {
  it(`MockComponent with HOC`, () => {
    const component = renderer
      .create(
          <MockComponentWrapped
            src={src}
            isPlaying={false}
            handlePlay={() => {}}
          />,
          options
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
