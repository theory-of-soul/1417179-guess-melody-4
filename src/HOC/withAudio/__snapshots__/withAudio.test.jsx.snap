// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` Hoc withAudio snapshot tests MockComponent with HOC 1`] = `
<div>
  <audio
    src=""
  />
</div>
`;
